Update rollover index API docs for data streams (#55551)
elastic/elasticsearch: docs/reference/indices/rollover-index.asciidoc

<titleabbrev>Rollover index</titleabbrev>
++++

> Creates a new index for a rollover target when the target's existing index meets a

Reviewer: 👍  I feel like 'rollover target' is a good term for this and it's precisely explained here.

condition you provide. A rollover target can be either an <<indices-aliases, index alias>>

> or a <<indices-datastreams, data stream>>. When targeting an alias, the alias is updated to

Reviewer: `indices-datastreams` isn't a valid anchor in the docs right now. This is causing the docs CI to fail.

> or a <<indices-datastreams, data stream>>. When targeting an alias, the alias is updated to

It looks like we don't have data stream docs at all atm. It might be good to add those before merging this.

Author: @jrodewig, I've added those in https://github.com/elastic/elasticsearch/pull/55557 if you would like to see them. I presume they will need to be merged first to create the appropriate data streams anchors.

Reviewer: 👍 Awesome. And yes--- they'll need to be merged in before you can reference the anchor here. You may have to update this branch after #55557 merges to get the docs CI to pass.

> reference the new index. When targeting a data stream, the new index becomes the data stream's

Reviewer: nit: I think "point" is a bit more clear here.
```suggestion
point to the new index. When targeting a data stream, the new index becomes the data stream's
```

==== {api-request-title}

`POST /<rollover-target>/_rollover/<target-index>`

`POST /<rollover-target>/_rollover/`

[[rollover-index-api-desc]]
==== {api-description-title}

> The rollover index API rolls an existing index for a rollover target to a new index when
> the existing index meets a condition you provide. You can use this API to retire
> an index that becomes too large or too old.

Reviewer: ```suggestion
The rollover index API rolls a rollover target's existing index to a new index
when the existing index meets a condition you provide. You can use this API to 
retire an index that becomes too large or too old.
```

automatically roll over indices when a condition is met, you can use {es}'s
<<index-lifecycle-management, index lifecycle management (ILM) policies>>.

The rollover index API accepts a rollover target name
and a list of `conditions`.

If the specified rollover target is an alias pointing to a single index,
the rollover request:

. Creates a new index
. Adds the alias to the new index
. Removes the alias from the original index

If the specified rollover target is an alias pointing to multiple indices,
one of these indices must have `is_write_index` set to `true`.
In this case, the rollover request:

. Creates a new index
. Sets `is_write_index` to `true` for the new index
. Sets `is_write_index` to `false` for the original index

If the specified rollover target is a data stream, the rollover request:

. Creates a new index

> . Adds the new index as a backing index on the data stream

Reviewer: Should we mention that it's also the write index?

[[rollover-index-api-path-params]]
==== {api-path-parms-title}

`<rollover-target>`::

> (Required, string)

Reviewer: For parameters that are required in some situations, we typically add an asterisk.
```suggestion
(Required*, string)
```

+
--
(Required, object)

> Set of conditions the rollover target's existing index must meet to roll over.

Reviewer: Good catch!

<2> Whether the rollover was dry run.
<3> The result of each condition.

[[rollover-data-stream-ex]]
===== Roll over a data stream

[source,console]
--------------------------------------------------

> PUT _data_stream/my-data-stream <1>

Reviewer: ```suggestion
PUT /_data_stream/my-data-stream <1>
```

{
  "timestamp_field": "@timestamp"
}

# Add > 1000 documents to my-data-stream

POST /my-data-stream/_rollover <2>
{
  "conditions" : {
  "max_age": "7d",
  "max_docs": 1000,
  "max_size": "5gb"
  }
}
--------------------------------------------------
<1> Creates a data stream called `my-data-stream` with one initial backing index
named `my-data-stream-000001`.

> <2> If the current index of `my-data-stream` was created 7 or more days ago, or
> contains 1,000 or more documents, or has an index size at least around 5GB, then
> a new backing index `my-data-stream-000002` index is created and added to the
> `my-data-stream` data stream.

Reviewer: The conditions become a bit hard to parse in a callout. I think breaking them out into a bulleted list might help>
```suggestion
<2> This request creates a new backing index, `my-data-stream-000002`, and adds
it as the write index for the `my-data-stream` data stream if the current
write index meets at least one of the following conditions:
+
--
* The index was created 7 or more days ago.
* The index has an index size of 5GB or greater.
* The index contains 1,000 or more documents.
--
```

[[rollover-index-specify-index-ex]]
===== Specify a target index name

> If the rollover target is a data stream, you may not specify a target index

Reviewer: nit
```suggestion
If the rollover target is a data stream, you cannot specify a target index
```
